File: snake.js
/*
 Original Code Credits: Chris DeLeon of HomeTeam GameDev 
 Modified by: Yakko Majuri
*/

'use strict'

let canvas, ctx;

window.onload = function () {
    canvas = document.getElementById("game-board");
    ctx = canvas.getContext("2d");
    document.addEventListener("keydown", keyPush);
    setInterval(game, 1000 / 15);
    let instructionsBtn = document.getElementById("instructions-btn");
    instructionsBtn.addEventListener("click", showInstructions)
}

let score = 0;
let highScore = 0;

let px = 10;
let py = 10;
let gs = 20;
let tc = 20;
let ax = 15;
let ay = 15;
let xv = 0;
let yv = 0;
let trail = [];
let tail = 5;

let paused = false
let pre_pause = [0, 0]



function game() {
    px += xv;
    py += yv;
    if (px < 0) {
        px = tc - 1;
    }
    if (px > tc - 1) {
        px = 0;
    }
    if (py < 0) {
        py = tc - 1;
    }
    if (py > tc - 1) {
        py = 0;
    }
    ctx.fillStyle = "black";
    ctx.fillRect(0, 0, canvas.width, canvas.height);

    ctx.fillStyle = "lime";
    for (var i = 0; i < trail.length; ++i) {
        ctx.fillRect(trail[i].x * gs, trail[i].y * gs, gs - 2, gs - 2);
        if (trail[i].x == px && trail[i].y == py) {
            tail = 5;

            score = paused ? score : 0;
        }
    }
    trail.push({ x: px, y: py });
    while (trail.length > tail) {
        trail.shift();
    }

    if (ax == px && ay == py) {
        ++score;
        ++tail;
        ax = Math.floor(Math.random() * tc);
        ay = Math.floor(Math.random() * tc);
    }
    ctx.fillStyle = "red";
    ctx.fillRect(ax * gs, ay * gs, gs - 2, gs - 2);
    updateScore(score)
}

function updateScore(newScore) {
    document.getElementById("score").innerHTML = "Score: " + String(newScore);
    if (newScore > highScore) {
        document.getElementById("high-score").innerHTML = "High Score: " + String(newScore);
        highScore = newScore;
    }
}

function keyPush(evt) {
    if (!paused) {
        switch (evt.keyCode) {
            case 37:
                xv = -1; yv = 0;
                break;
            case 38:
                xv = 0; yv = -1;
                break;
            case 39:
                xv = 1; yv = 0;
                break;
            case 40:
                xv = 0; yv = 1;
                break;
        }
    }
    if (evt.keyCode === 32) {
        let temp = [xv, yv];
        xv = pre_pause[0];
        yv = pre_pause[1];
        pre_pause = [...temp];
        paused = !paused;
    }
}

function showInstructions() {
    window.alert("Use the arrow keys to move and press spacebar to pause the game.")
}

File: snake.py
from browser import document, html, window
from javascript import Math

score = 0
high_score = 0
px = py = 10
gs = tc = 20
ax = ay = 15
xv = yv = 0
trail = []
tail = 5
pre_pause = [0,0]
paused = False

def game():
    global px, py, tc, gs, ax, ay, trail, tail, score
    px += xv
    py += yv
    if px < 0:
        px = tc-1
    if px > tc-1:
        px = 0
    if py < 0:
        py = tc-1
    if py > tc-1:
        py = 0
    ctx.fillStyle = "black"
    ctx.fillRect(0, 0, canvas.width, canvas.height)
    ctx.fillStyle = "lime"
    for i in range(len(trail)):
        ctx.fillRect(trail[i][0]*gs, trail[i][1]*gs, gs-2, gs-2)
        if trail[i][0] == px and trail[i][1] == py:
            score = score if paused else 0 
            tail = 5
    trail.insert(0, [px, py])
    while len(trail) > tail:
        trail.pop()

    if ax == px and ay == py:
        tail += 1
        ax = Math.floor(Math.random()*tc)
        ay = Math.floor(Math.random()*tc)
        score += 1
    update_score(score)
    ctx.fillStyle = "red"
    ctx.fillRect(ax*gs, ay*gs, gs-2, gs-2)

def update_score(new_score):
    global high_score
    document["score"].innerHTML = "Score: " + str(new_score)
    if new_score > high_score:
        document["high-score"].innerHTML = "High Score: " + str(new_score)
        high_score = new_score
        
def key_push(evt):
    global xv, yv, pre_pause, paused
    key = evt.keyCode
    if key == 37 and not paused:
        xv = -1
        yv = 0
    elif key == 38 and not paused:
        xv = 0
        yv = -1
    elif key == 39 and not paused:
        xv = 1
        yv = 0
    elif key == 40 and not paused:
        xv = 0
        yv = 1
    elif key == 32:
        temp = [xv, yv]
        xv = pre_pause[0]
        yv = pre_pause[1]
        pre_pause = [*temp]
        paused = not paused
        
    
def show_instructions(evt):
    window.alert("Use the arrow keys to move and press spacebar to pause the game.")

canvas = document["game-board"]
ctx = canvas.getContext("2d")
document.addEventListener("keydown", key_push)
game_loop = window.setInterval(game, 1000/10)
instructions_btn = document["instructions-btn"]
instructions_btn.addEventListener("click", show_instructions)